clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -1536,6 +1539,14 @@ void CodeGenModule::GenOpenCLArgMetadata(llvm::Function *Fn,
 
       // Get argument name.
       argNames.push_back(llvm::MDString::get(VMContext, parm->getName()));
+
+      auto *SYCLBufferLocationAttr =
```

Comment:
Why doesn't 'no metadata' mean 'not set'?  Why is the -1 value necessary?

Reply:
I don't quiet get the question, but I think that I shall state it more explicitly: the idea is to add the attribute only for pointer kernel parameters, that were generated for accessors  annotated with the property, hence not every parameter has the attribute applied (even the accessor pointer can has no such attribute without the property set). '-1' here means, that compiler's backend shan't make assumptions on where exactly the buffer was allocated and shall follow a default protocol for load-store units resolution. 

And there is one more thing that is needed to be clarified. Number of metadata values equals number of (not only pointer) kernel parameters - it's an implementation detail. It historically happened that with such metadata we annotates every kernel parameter, so '-1' for non-pointer kernel parameters just exists and means nothing. I'll look through the code and find where such a comment fits the best.

Reply:
Hmm... I still find myself wondering, if the backend is intending to ignore the -1 metadata, why bother putting it there at all?  You say 'number of metadata values equals number of kernel parmeters', but is that necessarily true?

Basically my question is: Can we avoid making the -1 value and just replace it with 'no data' to mean the same thing?  It doesn't seem that there is a differentiation between the two.

Reply:
The value -1 itself is not important. As long as we can figure out, from the metadata, that the buffer_location for that particular argument is empty, we're okay. For OpenCL, that used to be an empty string `!""`. But it is important to have the number of metadata values equal to the number of kernel parameters to avoid any ambiguity (keep the 1:1 mapping).

Reply:
Well, that was the original requirement to make '-1' to be the default value, I personally don't mind to change it (would need to make a change in the translator though).

Reply:
I'm personally okay with -1 and would rather not have us change the translator now since that change is already in. We could always change that later I suppose if we think there's a better option.

Reply:
It seems like a waste to keep the -1s, but at least this isn't wasting space in the AST.  Resolve for now.